File: Popper/src/PopperHelper/PopperHelper.js
// @ts-check

import { createPopper } from "../node_modules/@popperjs/core/dist/esm/popper.js";

import { generateId } from "./DomHelpers";

export class PopperHelper {
    static cache = {};

    static construct(containerRef, popperRef, options) {
        const id = generateId();

        const popper = createPopper(containerRef, popperRef, options);

        PopperHelper.cache[id] = {
            id,
            popper,
        };

        setTimeout(() => popper.update());

        return id;
    }

    static dispose(id) {
        const record = PopperHelper.cache[id];
        if (record && record.popper) {
            record.popper.destroy();
        }
        delete PopperHelper.cache[id];
    }

    static update(id) {
        const record = PopperHelper.cache[id];
        if (record && record.popper) {
            record.popper.update();
        }
    }
}

File: Popper/src/PopperHelper/PopperHelper.cs
using System.Threading.Tasks;
using System.Collections.Generic;
using Microsoft.AspNetCore.Components;
using Skclusive.Core.Component;
using Skclusive.Script.DomHelpers;
using System;

namespace Skclusive.Material.Popper
{
    public class PopperHelper : IAsyncDisposable
    #if NETSTANDARD2_0
        , IDisposable
    #endif
    {
        public PopperHelper(IScriptService scriptService)
        {
            ScriptService = scriptService;
        }

        private object Id;

        private IScriptService ScriptService { get; }

        public async ValueTask InitAsync(ElementReference? containerRef, ElementReference? popperRef, PopperOptions options)
        {
            Id = await ScriptService.InvokeAsync<object>("Skclusive.Material.Popper.PopperHelper.construct", containerRef, popperRef, options);
        }

        public async ValueTask UpdateAsync()
        {
            if (Id != null)
            {
                await ScriptService.InvokeVoidAsync("Skclusive.Material.Popper.PopperHelper.update", Id);
            }
        }

        public async ValueTask DisposeAsync()
        {
            if (Id != null)
            {
                await ScriptService.InvokeVoidAsync("Skclusive.Material.Popper.PopperHelper.dispose", Id);

                Id = null;
            }
        }

        public bool Active => Id != null;

#if NETSTANDARD2_0

        void IDisposable.Dispose()
        {
            if (this is IAsyncDisposable disposable)
            {
                _ = disposable.DisposeAsync();
            }
        }

#endif
    }
}